lib/platform/linux-uart.c

```diff
@@ -0,0 +1,423 @@
+/*
+ * Microsemi Switchtec(tm) PCIe Management Library
+ * Copyright (c) 2017, Microsemi Corporation
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ *
+ */
+
+#ifdef __linux__
+
+#include "../switchtec_priv.h"
+#include "switchtec/switchtec.h"
+#include "gasops.h"
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <errno.h>
+#include <stddef.h>
+#include <assert.h>
+#include <string.h>
+
+#include <sys/file.h>
+#include <termios.h>
+#include "../crc8.h"
+
+static const char *dbginfo_off = "pscdbg 0 all\r";
+static const char *echo_off = "echo 0\r";
+static const char *echo_on = "echo 1\r";
+
+struct switchtec_uart{
+	struct switchtec_dev dev;
+	int fd;
+};
+
+#define to_switchtec_uart(d) \
+	((struct switchtec_uart *) \
+	 ((char *)d - offsetof(struct switchtec_uart, dev)))
+
+
+#define UART_MAX_WRITE_BYTES			(100)
+#define RETRY_NUM				(1000)
+
+
+static int read_resp_line(int fd, char *str)
+{
+	int ret;
+	int i;
+	int cnt = 0;
+
+	for (i = 0; i < RETRY_NUM; i++) {
+
+		ret = read(fd, str + cnt, sizeof(str));
+		if (ret < 0)
+			return ret;
+	
+		cnt += ret;
+		str[cnt] = '\0';
+	
+		if (strrchr(str, '>') && strrchr(str, ':') &&
+			(strrchr(str, ':') + 5 == strrchr(str, '>'))){
+			return 0;
```

Comment:
I think a scanf here would be clearer... At the very least calling identical strrchr()s twice is wasteful...